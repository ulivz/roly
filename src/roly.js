import fs from 'fs'
import * as rollup from 'rollup'
import switchy from 'switchy'
import chalk from 'chalk'
import getRollupOptions from './get-rollup-options'
import log from './log'
import {
  cwd,
  isString,
  isArray,
  splitStrByComma,
  handleRollupError,
  joinPath,
  isAbsolutePath
} from './utils'

function readInPkg(file) {
  try {
    const pkg = require(file)
    const roly = pkg.roly || {}
    if (pkg.name && !roly.filename) {
      const { name } = pkg
      roly.filename = name.startsWith('@')
        ? name.slice(name.lastIndexOf('/') + 1)
        : name
    }
    roly.pkg = pkg
    return roly
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      return {
        // Ensure pkg in options
        pkg: {}
      }
    }
    throw err
  }
}

function singleRoly(options = {}) {
  return new Promise(resolve => {
    options = Object.assign(
      {
        input: './src/index.js',
        format: ['cjs'],
        outDir: './dist',
        filename: 'index',
        compress: []
      },
      options
    )

    // for backward-compat
    if (options.entry) {
      options.input = options.entry
    }

    // for custom base dir
    const baseDir = options.baseDir
    if (baseDir) {
      if (!isAbsolutePath(options.input)) {
        options.input = joinPath(baseDir, options.input)
      }
      if (!isAbsolutePath(options.outDir)) {
        options.outDir = joinPath(baseDir, options.outDir)
      }
    }

    // Ensure format is an array
    let formats = options.format
    if (isString(formats)) {
      formats = formats === 'all'
        ? ['cjs', 'umd', 'es']
        : splitStrByComma(formats)
    } else if (!isArray(formats)) {
      throw new TypeError('Expect "format" to be a string or Array')
    }

    // Ensure compress is an array
    if (isString(options.compress)) {
      options.compress = splitStrByComma(options.compress)
    } else if (options.compress === true) {
      options.compress = ['cjs', 'umd', 'iife'] // Currently uglifyjs cannot compress es format
    } else if (!isArray(options.compress)) {
      throw new TypeError('Expect "compress" to be a string/true or Array')
    }

    options.compress = options.compress
      .filter(v => formats.includes(v))
      .map(v => `${v}Compress`)

    const allFormats = [...formats, ...options.compress]
    resolve(allFormats)
  }).then(allFormats => {
    return Promise.all(
      allFormats.map(format => {
        const rollupOptions = getRollupOptions(options, format)

        if (options.watch) {
          let init
          const watcher = rollup.watch(rollupOptions)
          return watcher.on('event', event => {
            switchy({
              START() {
                log(format, 'starting', chalk.blue)
                if (!init) {
                  init = true
                }
              },
              BUNDLE_START() {},
              BUNDLE_END() {
                log(format, 'bundled', chalk.green)
              },
              END() {},
              ERROR() {
                handleRollupError(event.error)
              },
              FATAL() {
                handleRollupError(event.error)
              },
              default() {
                console.error('unknown event', event)
              }
            })(event.code)
          })
        }

        return rollup.rollup(rollupOptions).then(bundle => {
          if (options.write === false)
            return bundle.generate(rollupOptions.output)
          return bundle.write(rollupOptions.output)
        })
      })
    ).then(result => {
      return result.reduce((res, next, i) => {
        res[allFormats[i]] = next
        return res
      }, {})
    })
  })
}

export default function roly(options = {}) {
  const baseDir = options.baseDir || ''
  const configFileName = options.config || 'roly.config.js'
  const configFilePath = cwd(joinPath(baseDir, configFileName))

  let rolyConfig = {}
  if (fs.existsSync(configFilePath)) {
    rolyConfig = require(configFilePath)
  }

  const pkgConfig = readInPkg(cwd(joinPath(baseDir, 'package.json')))

  if (Array.isArray(rolyConfig)) {
    return Promise.all(
      rolyConfig.map(rolyOptions => {
        return singleRoly(Object.assign(options, pkgConfig, rolyOptions))
      })
    )
  }
  return singleRoly(Object.assign(options, rolyConfig, pkgConfig))
}
