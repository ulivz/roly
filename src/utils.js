/* eslint-disable import/prefer-default-export */
import path from 'path'
import chalk from 'chalk'
import log from './log'

/**
 * Array type check
 */
export function isArray(arr) {
  return Array.isArray(arr)
}

/**
 * String type check
 */
export function isString(str) {
  return typeof str === 'string'
}

/**
 * Shorthand to get the abosulte path relative
 * to current process working directory
 */
export function cwd(...args) {
  return path.resolve(process.cwd(), ...args)
}

/**
 * String type check
 */
export function splitStrByComma(str) {
  return str.split(',').map(item => item.trim())
}

/**
 * path
 */
export const joinPath = path.join
export const isAbsolutePath = path.isAbsolute

/**
 * Handle rollup error
 */
export function handleRollupError(error) {
  console.log('123213')
  log(error.plugin || 'error', error.message, chalk.red)
  if (error.id) {
    console.log(chalk.dim(`Location: ${error.id}`))
  }
  console.error(error.snippet || error)
  console.log()
  process.exitCode = 1
}
