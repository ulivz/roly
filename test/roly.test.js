import path from 'path'
import fs from 'fs-extra'
import switchy from 'switchy'
import {EventEmitter} from 'events'
import {cwd} from './utils'
import roly from '../src/roly'

const prevCwd = process.cwd()

beforeAll(() => {
  process.chdir(__dirname)
})

afterAll(() => {
  process.chdir(prevCwd)
})

describe('roly', () => {

  test('it throws because entry not found', () => {
    return roly().catch(err => {
      expect(err.message).toEqual('Could not resolve entry (./src/index.js)')
    })
  })

  /**
   * When package.json's name === @name/package-name, then module's file name === package-name
   */
  test('it should remove the name prefix from a scoped package name', async () => {
    await roly({
      baseDir: 'fixtures/scoped',
      entry: cwd('fixtures/entry.js'),
      exports: 'named',
      outDir: 'dist'
    })
    const files = await fs.readdir(cwd(`fixtures/scoped/dist`))
    expect(files).toEqual(['package-name.common.js'])
  })

  /**
   * When package's name and roly.filename exist at the same time, use roly.filename
   */
  test('it should use the roly.filename', async () => {
    await roly({
      baseDir: 'fixtures/filename',
      entry: cwd('fixtures/entry.js'),
      exports: 'named',
      outDir: 'dist'
    })
    const files = await fs.readdir(cwd(`fixtures/filename/dist`))
    expect(files).toEqual(['roly.common.js'])
  })

  test('should generate all bundles', async () => {
    const result = await roly({
      entry: 'fixtures/entry.js',
      format: 'all',
      exports: 'named',
      write: false
    })
    expect(Object.keys(result)).toHaveLength(3)
  })

  test('it replaces string using rollup-plugin-replace', async () => {
    const { cjs } = await roly({
      entry: 'fixtures/entry.js',
      exports: 'named',
      replace: {
        __VERSION__: '0.0.0'
      },
      write: false
    })
    expect(cjs.code).toMatchSnapshot()
  })

  test('should throw error when "format" is neither string nor array', () => {
    return roly({ format: null }).catch(err => {
      expect(err.message).toEqual('Expect "format" to be a string or Array')
    })
  })

  test('should throw error when "compress" is neither string/true nor array', () => {
    return roly({ compress: null }).catch(err => {
      expect(err.message).toEqual('Expect "compress" to be a string/true or Array')
    })
  })

  test('should work on watch mode', async () => {
    const watchers = await roly({
      entry: 'fixtures/entry.js',
      exports: 'named',
      watch: true
    })
    Object.keys(watchers).forEach(format => {
      const watcher = watchers[format]
      expect(watcher instanceof EventEmitter).toBeTruthy()
      watcher.on('event', event => {
        switchy({
          ERROR() {
            expect(process.exitCode).toEqual(1)
            process.exitCode = 0
          },
          FATAL() {
            expect(process.exitCode).toEqual(1)
            process.exitCode = 0
          }
        })(event.code)
      })
      watcher.emit('event', { code: 'START' })
      watcher.emit('event', { code: 'BUNDLE_END' })

      watcher.emit('event', { code: 'ERROR', error: 'error' })
      watcher.emit('event', { code: 'FATAL', error: 'error' })
      watcher.emit('event', { code: 'ULIVZ' })
      setImmediate(() => {
        watcher.close()
      })
    })
  })
})


describe('option - banner', () => {
  test('banner option = Boolean', async () => {
    const es = await roly({
      baseDir: 'fixtures/banner', // custom the working directory
      entry: cwd('fixtures/entry.js'),
      format: 'es',
      exports: 'named',
      banner: true // banner info from package.json
    })
    expect(es.code).toMatchSnapshot()

  })

  test('banner option = Object', async () => {
    const { cjs } = await roly({
      entry: 'fixtures/entry.js',
      format: 'cjs',
      exports: 'named',
      banner: {
        name: 'roly',
        version: '5.2.0',
        author: 'ulivz',
        license: 'MIT'
      },
      write: false
    })
    expect(cjs.code).toMatchSnapshot()
  })

  test('banner option = String', async () => {
    const { umd, umdCompress } = await roly({
      entry: 'fixtures/entry.js',
      format: 'umd',
      compress: true,
      exports: 'named',
      write: false,
      banner: '/*! rolyroly */'
    })
    expect(umd.code).toMatchSnapshot()
    expect(umdCompress.code).toMatchSnapshot()
  })

})