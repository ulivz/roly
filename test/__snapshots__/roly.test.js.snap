// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`option - banner banner option = Boolean 1`] = `undefined`;

exports[`option - banner banner option = Object 1`] = `
"/*!
 * roly v5.2.0
 * (c) 2017-present ulivz
 * Released under the MIT License.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var a = 1;
var version = '__VERSION__';

exports['default'] = a;
exports.version = version;
"
`;

exports[`option - banner banner option = String 1`] = `
"/*! rolyroly */
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.index = {})));
}(this, (function (exports) { 'use strict';

var a = 1;
var version = '__VERSION__';

exports['default'] = a;
exports.version = version;

Object.defineProperty(exports, '__esModule', { value: true });

})));
"
`;

exports[`option - banner banner option = String 2`] = `
"/*! rolyroly */
!function(e,t){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?t(exports):\\"function\\"==typeof define&&define.amd?define([\\"exports\\"],t):t(e.index={})}(this,function(e){\\"use strict\\";e.default=1,e.version=\\"__VERSION__\\",Object.defineProperty(e,\\"__esModule\\",{value:!0})});
"
`;

exports[`roly it replaces string using rollup-plugin-replace 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var a = 1;
var version = '0.0.0';

exports['default'] = a;
exports.version = version;
"
`;
